src/avx512bw/avx512bw-from-utf8.inl.cpp

```diff
@@ -0,0 +1,466 @@
+
+    c = (continuation1 & 0x3f)
+    x = (c ^ fixup) > 0x1f
+
+    where `fixup` is a function of nibble value:
+
+        fixup[nibble] = 0  => 0x00
+                        14 => 0x40
+                        _  => 0xc0
+
+    case 1 fixup = 0x00:
+                                                          c ^ fixup > 0x1f
+    c <= 0x1f: 0b000x_xxxx ^ 0xb0000_0000 = 0xb000x_xxxx  false
+    c >  0x1f: 0b001x_xxxx ^ 0xb0000_0000 = 0xb001x_xxxx  true
+
+    case 2 fixup = 0x40:
+                                                          c ^ fixup > 0x1f
+    c <= 0x1f: 0b000x_xxxx ^ 0xb0010_0000 = 0xb001x_xxxx  true
+    c >  0x1f: 0b001x_xxxx ^ 0xb0010_0000 = 0xb000x_xxxx  false
+
+    case 3 fixup = 0x40:
+                                                          c ^ fixup > 0x1f
+    c <= 0x1f: 0b000x_xxxx ^ 0xb1100_0000 = 0xb110x_xxxx  true
+    c >  0x1f: 0b001x_xxxx ^ 0xb1100_0000 = 0xb111x_xxxx  true
+
+
+c) For 4-byte chars we decide based on the lower nibble of the leading byte:
+
+- 0: valid if (continuation1 & 0x3f) > 0x0f
+- 1, 2, 3: always valid
+- 4: valid if (continuation1 & 0x3f) <= 0x0f [negation of case #0]
+- 5, 15: always invalid
+
+In this case we build a 2-bit values where one bit stores result of
+(continuation1 & 0x3f) > 0x0f and another is negation of that bit.  Then case
+#0 is encoded as mask 0b01, case #4 is encoded as 0b10, cases #{1,2,3} as 0b11
+and the remaining cases as 0b00. The validity of chars is checked with the
+following expression:
+
+    value & mask != 0
+
+To create a 2-bit value we subtract from c = (continuation1 & 0x3f) the number
+0x10. If c = 0b0000_xxxx (<= 0x0f) then the result of the subtraction is always
+in form 0b1111_xxxx. The two most significant bits encodes condition (c <= 0x0f).
+Now we negate 6th bit, and then the these two most significant bits represents
+the required combination which is later tested.
+*/
+__mmask64 validate_leading_bytes(__m512i leading_bytes, __m512i continuation1, __mmask64 tested_chars) {
+
+    const __m512i nibble0 = _mm512_and_si512(leading_bytes, v_0f);
+
+    // 1. Assume all non-continuation bytes are valid leading bytes
+    //    We unset all continuation bytes (0b10xx_xxxx) and ASCII chars
+    //    (0b0xxx_xxxx) --- so we looking for 0b11xx_xxxx chars.
+    //
+    //    (We may mask-out some leading bytes via `tested_chars`).
+    __mmask64 valid;
+    {
+        const __m512i t0 = _mm512_and_si512(leading_bytes, v_c0);
+        valid = _mm512_cmpeq_epi8_mask(t0, v_c0);
+    }
+
+    __mmask64 _2bytes;
+    {
+        const __m512i t0 = _mm512_and_si512(leading_bytes, v_e0);
+        _2bytes = _mm512_cmpeq_epi8_mask(t0, v_c0);
+    }
+
+    __mmask64 _3bytes;
+    {
+        const __m512i t0 = _mm512_and_si512(leading_bytes, v_f0);
+        _3bytes = _mm512_cmpeq_epi8_mask(t0, v_e0);
+    }
+
+    __mmask64 _4bytes;
+    {
+        const __m512i t0 = _mm512_and_si512(leading_bytes, v_f8);
+        _4bytes = _mm512_cmpeq_epi8_mask(t0, v_f0);
```

Comment:
Same kind of remark, this can be done with a single instruction... maybe _mm512_cmpgt_epu8_mask